L70-node-proto-loader-type-generator.md

```diff
@@ -0,0 +1,67 @@
+@grpc/proto-loader TypeScript Type Generator CLI Tool
+----
+* Author(s): murgatroid99
+* Approver: wenbozhu
+* Status: Draft
+* Implemented in: Node.js
+* Last updated: 2020-06-01
+* Discussion at: https://groups.google.com/g/grpc-io/c/FwLprMC-5Qc
+
+## Abstract
+
+Add a tool to the `@grpc/proto-loader` to generate types that describe the objects that will be generated by loading the output of `@grpc/proto-loader` into `grpc` or `@grpc/grpc-js`.
+
+## Background
+
+`@grpc/proto-loader` outputs objects with types determined by the `.proto` files loaded at runtime, so it is currently very difficult to get compile-time type information about those objects. As TypeScript becomes increasingly popular, that compile-time type information becomes increasingly desirable.
+
+
+### Related Proposals: 
+* L23 Standalone Node.js gRPC+Protobuf.js Library API
+
+## Proposal
+
+In the `@grpc/proto-loader` library, provide a command line tool that will generate the types of the objects that will be created by passing the output of `load` or `loadSync` to `grpc.loadPackageDefinition` when loading a specific set of `.proto` files with a specific set of options. This will allow users to write code like this to use the type information in the rest of the code:
+
+```ts
+import * as grpc from 'grpc';
+// OR 
+import * as grpc from '@grpc/grpc-js';
+
+import ProtoGrpcType from 'path/to/proto-file-name'; // File generated by the tool is path/to/proto-file-name.d.ts
+import * as protoLoader from '@grpc/proto-loader';
+
+const packageDefinition = protoLoader.loadSync('other/path/to/proto-file-name.proto', options);
+const loadedPackageDefinition = grpc.loadPackageDefinition(packageDefinition) as ProtoGrpcType<grpc.Client, grpc.CallOptions>; // How the types will be used
+```
+
+The tool will be named `proto-loader-gen-types` and will have this usage:
+
+```
+proto-loader-gen-types [OPTIONS] <proto file name>
+```
+
+The options will correspond directly to the `protoLoader.load` options as follows:
+
+ - `--keep-case`: Preserve the case of field names
+ - `--longs=String|Number`: Specify the type that should be used to output 64 bit integer values
+ - `--enums=String`: Specify that enum values should be output as strings
+ - `--bytes=Array|String`: Specify the type that should be used to output `bytes` fields
+ - `--defaults`: Indicates that default values should be output for missing fields
+ - `--arrays`: Indicates that empty arrays should be output for missing repeated fields even if `--defaults` is unset
+ - `--objects`: Indicates that empty objects should be output for missing message fields even if `--defaults` is unset
+ - `--oneofs`: Indicates that virtual "oneof" fields will be set to the present field's name in the output
+ - `--includeDirs=<directories...>`, `-I`: A list of directories to search for included `.proto` files
+
+ - `--outDir`, `-O`: The path in which to output the `.d.ts` file

```

Comment:
if the output is a single `.d.ts` file, would it be better to have a regular `-o` option for an output file name, dumping it to `stdout` by default if the path is not given?

Reply:
if it's not a single `.d.ts` but multiple `.d.ts` (e.g. one per service), it's worth mentioning that too.

Reply:
I've clarified what files will be generated and in what directory structure.